File: extras/Python/cmd_messenger/received_command.py
"""``ReceivedCommand`` — port of C# ``ReceivedCommand``.

Provides two reader APIs (decision 7b):

1. **Iterator-style** (C# parity): ``read_int32_arg()``, ``read_float_arg()``,
   ``read_string_arg()``, ``read_bool_arg()``, etc. — and the binary variants
   ``read_bin_*_arg()``. Each call advances the internal parameter cursor.

2. **Format-string** (Pythonic): :meth:`read` parses a format string and
   returns a tuple. Codes mirror :mod:`struct` where sensible:

   =====  =========================  ==============================
   code   text type                  binary variant (``*`` prefix)
   =====  =========================  ==============================
   ``i``  int32                      ``*i``
   ``I``  uint32                     ``*I``
   ``h``  int16                      ``*h``
   ``H``  uint16                     ``*H``
   ``b``  byte (signed)              ``*b``
   ``B``  byte (unsigned)            ``*B``
   ``f``  float (single)             ``*f``
   ``d``  double (BoardType-aware)   ``*d``
   ``s``  string                     ``*s`` (escape-decoded bytes)
   ``?``  bool                       ``*?``
   ``c``  single char                ``*c``
   =====  =========================  ==============================

   Trailing ``*`` repeats the previous code for all remaining args:
   ``read("f*")`` reads every remaining argument as a float.
"""
from __future__ import annotations

from typing import Any, List, Optional, Tuple, Union

from . import binary_converter, escaping
from .command import Command
from .enums import BoardType


class ReceivedCommand(Command):
    """A command received from the Arduino."""

    def __init__(self, raw_arguments: Optional[List[str]] = None) -> None:
        super().__init__()
        self.raw_string: str = ""
        self._parameter: int = -1
        self._dumped: bool = True

        if not raw_arguments:
            return

        try:
            self.cmd_id = int(raw_arguments[0])
        except (ValueError, IndexError):
            self.cmd_id = -1
            return

        if self.cmd_id < 0:
            return

        if len(raw_arguments) > 1:
            self._cmd_args.extend(raw_arguments[1:])

    # ------------------------------------------------------------------
    # Iterator
    # ------------------------------------------------------------------
    def next(self) -> bool:
        """Advance to the next argument. Returns ``False`` when exhausted."""
        if self._dumped:
            if self._parameter < len(self._cmd_args) - 1:
                self._parameter += 1
                self._dumped = False
                return True
            return False
        return True

    def available(self) -> bool:
        """Return whether another argument is available (alias of :meth:`next`)."""
        return self.next()

    def __iter__(self):
        """Iterate over the raw string arguments."""
        return iter(list(self._cmd_args))

    # ------------------------------------------------------------------
    # Text-mode typed readers (mirror C# ReadXxxArg)
    # ------------------------------------------------------------------
    def _current(self) -> str:
        return self._cmd_args[self._parameter]

    def read_int16_arg(self) -> int:
        if self.next():
            try:
                v = int(self._current())
                # Range check to mimic C# Int16.TryParse failure -> 0.
                if -32768 <= v <= 32767:
                    self._dumped = True
                    return v
            except ValueError:
                pass
        return 0

    def read_uint16_arg(self) -> int:
        if self.next():
            try:
                v = int(self._current())
                if 0 <= v <= 65535:
                    self._dumped = True
                    return v
            except ValueError:
                pass
        return 0

    def read_int32_arg(self) -> int:
        if self.next():
            try:
                v = int(self._current())
                if -(2 ** 31) <= v <= (2 ** 31) - 1:
                    self._dumped = True
                    return v
            except ValueError:
                pass
        return 0

    def read_uint32_arg(self) -> int:
        if self.next():
            try:
                v = int(self._current())
                if 0 <= v <= (2 ** 32) - 1:
                    self._dumped = True
                    return v
            except ValueError:
                pass
        return 0

    def read_float_arg(self) -> float:
        if self.next():
            try:
                v = float(self._current())
                self._dumped = True
                return v
            except ValueError:
                pass
        return 0.0

    def read_double_arg(self) -> float:
        # Matches C#: requires comm manager because Bit16 demotes to float.
        if self.communication_manager is None:
            raise RuntimeError("CommunicationManager was not set for command.")
        if self.next():
            try:
                v = float(self._current())
                self._dumped = True
                return v
            except ValueError:
                pass
        return 0.0

    def read_string_arg(self) -> str:
        if self.next():
            v = self._current()
            if v is not None:
                self._dumped = True
                return v
        return ""

    def read_bool_arg(self) -> bool:
        return self.read_int32_arg() != 0

    def read_char_arg(self) -> str:
        s = self.read_string_arg()
        return s[0] if s else "\0"

    # ------------------------------------------------------------------
    # Binary-mode readers (mirror C# ReadBinXxxArg)
    # ------------------------------------------------------------------
    def read_bin_int16_arg(self) -> int:
        if self.next():
            v = binary_converter.to_int16(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0

    def read_bin_uint16_arg(self) -> int:
        if self.next():
            v = binary_converter.to_uint16(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0

    def read_bin_int32_arg(self) -> int:
        if self.next():
            v = binary_converter.to_int32(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0

    def read_bin_uint32_arg(self) -> int:
        if self.next():
            v = binary_converter.to_uint32(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0

    def read_bin_float_arg(self) -> float:
        if self.next():
            v = binary_converter.to_float(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0.0

    def read_bin_double_arg(self) -> float:
        if self.communication_manager is None:
            raise RuntimeError("CommunicationManager was not set for command.")
        if self.next():
            if self.communication_manager.board_type == BoardType.BIT_16:
                v: Optional[float] = binary_converter.to_float(self._current())
            else:
                v = binary_converter.to_double(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0.0

    def read_bin_byte_arg(self) -> int:
        if self.next():
            v = binary_converter.to_byte(self._current())
            if v is not None:
                self._dumped = True
                return v
        return 0

    def read_bin_string_arg(self) -> str:
        if self.next():
            b = binary_converter.escaped_string_to_bytes(self._current())
            if b is not None:
                self._dumped = True
                return b.decode(binary_converter.get_string_encoding())
        return ""

    def read_bin_bool_arg(self) -> bool:
        return self.read_bin_byte_arg() != 0

    # ------------------------------------------------------------------
    # Format-string reader (Pythonic extra — decision 7b)
    # ------------------------------------------------------------------
    _TEXT_READERS = {
        "i": "read_int32_arg",
        "I": "read_uint32_arg",
        "h": "read_int16_arg",
        "H": "read_uint16_arg",
        "b": "read_int16_arg",   # signed byte → narrow int16 reader is fine
        "B": "read_uint16_arg",  # unsigned byte
        "f": "read_float_arg",
        "d": "read_double_arg",
        "s": "read_string_arg",
        "?": "read_bool_arg",
        "c": "read_char_arg",
    }
    _BIN_READERS = {
        "i": "read_bin_int32_arg",
        "I": "read_bin_uint32_arg",
        "h": "read_bin_int16_arg",
        "H": "read_bin_uint16_arg",
        "b": "read_bin_byte_arg",
        "B": "read_bin_byte_arg",
        "f": "read_bin_float_arg",
        "d": "read_bin_double_arg",
        "s": "read_bin_string_arg",
        "?": "read_bin_bool_arg",
    }

    def read(self, fmt: str) -> Union[Any, Tuple[Any, ...]]:
        """Read arguments according to a format string. See module docstring.

        Returns a single value when ``fmt`` resolves to one read, otherwise a
        tuple — matching PyCmdMessenger ergonomics.
        """
        tokens = self._parse_format(fmt)
        results: List[Any] = []
        for is_binary, code in tokens:
            table = self._BIN_READERS if is_binary else self._TEXT_READERS
            reader_name = table.get(code)
            if reader_name is None:
                raise ValueError(f"Unknown format code {'*' if is_binary else ''}{code!r}")
            results.append(getattr(self, reader_name)())
        if len(results) == 1:
            return results[0]
        return tuple(results)

    def _parse_format(self, fmt: str) -> List[Tuple[bool, str]]:
        """Expand ``fmt`` to a list of ``(is_binary, code)`` tokens."""
        tokens: List[Tuple[bool, str]] = []
        i = 0
        n = len(fmt)
        last: Optional[Tuple[bool, str]] = None
        while i < n:
            ch = fmt[i]
            if ch == "*":
                # Trailing '*' alone repeats the previous code for all
                # remaining arguments not yet consumed by earlier tokens.
                if i == n - 1:
                    if last is None:
                        raise ValueError(
                            "Trailing '*' has no preceding format code to repeat."
                        )
                    # Args remaining at read() entry, accounting for cursor + dumped flag.
                    remaining = len(self._cmd_args) - (
                        self._parameter + (1 if self._dumped else 0)
                    )
                    remaining -= len(tokens)
                    for _ in range(max(remaining, 0)):
                        tokens.append(last)
                    i += 1
                    continue
                # Otherwise '*X' = binary code.
                if i + 1 >= n:
                    raise ValueError("'*' must be followed by a format code.")
                code = fmt[i + 1]
                token = (True, code)
                tokens.append(token)
                last = token
                i += 2
            else:
                token = (False, ch)
                tokens.append(token)
                last = token
                i += 1
        return tokens

File: extras/TypeScript/src/receivedCommand.ts
import { BinaryConverter } from './binaryConverter.js';
import { Command } from './command.js';
import { latin1Decode } from './encoding.js';
import { BoardType } from './enums.js';

type FormatToken = { binary: boolean; code: string };

function parseInteger(value: string, min: number, max: number): number | null {
  const trimmed = value.trim();
  if (!/^[+-]?\d+$/.test(trimmed)) {
    return null;
  }
  const parsed = Number(trimmed);
  if (!Number.isInteger(parsed) || parsed < min || parsed > max) {
    return null;
  }
  return parsed;
}

function parseFloatText(value: string): number | null {
  const trimmed = value.trim();
  if (trimmed === '') {
    return null;
  }
  const parsed = Number(trimmed);
  return Number.isFinite(parsed) ? parsed : null;
}

export class ReceivedCommand extends Command implements Iterable<string> {
  rawString = '';
  private parameter = -1;
  private dumped = true;

  constructor(rawArguments?: readonly string[]) {
    super();
    if (rawArguments === undefined || rawArguments.length === 0) {
      return;
    }

    const cmdId = parseInteger(rawArguments[0] ?? '', 0, Number.MAX_SAFE_INTEGER);
    if (cmdId === null) {
      this.cmdId = -1;
      return;
    }

    this.cmdId = cmdId;
    this.cmdArgs.push(...rawArguments.slice(1));
  }

  next(): boolean {
    if (this.dumped) {
      if (this.parameter < this.cmdArgs.length - 1) {
        this.parameter += 1;
        this.dumped = false;
        return true;
      }
      return false;
    }
    return true;
  }

  available(): boolean {
    return this.next();
  }

  [Symbol.iterator](): Iterator<string> {
    return this.arguments[Symbol.iterator]();
  }

  readInt16Arg(): number {
    return this.readInteger(-32768, 32767);
  }

  readUint16Arg(): number {
    return this.readInteger(0, 65535);
  }

  readInt32Arg(): number {
    return this.readInteger(-(2 ** 31), 2 ** 31 - 1);
  }

  readUint32Arg(): number {
    return this.readInteger(0, 2 ** 32 - 1);
  }

  readFloatArg(): number {
    if (this.next()) {
      const parsed = parseFloatText(this.current());
      if (parsed !== null) {
        this.dumped = true;
        return parsed;
      }
    }
    return 0;
  }

  readDoubleArg(): number {
    if (this.communicationManager === undefined) {
      throw new Error('CommunicationManager was not set for command.');
    }
    return this.readFloatArg();
  }

  readStringArg(): string {
    if (this.next()) {
      this.dumped = true;
      return this.current();
    }
    return '';
  }

  readBoolArg(): boolean {
    return this.readInt32Arg() !== 0;
  }

  readCharArg(): string {
    const value = this.readStringArg();
    return value.length > 0 ? value[0] ?? '\0' : '\0';
  }

  readBinInt16Arg(): number {
    return this.readBinaryNumber((value) => BinaryConverter.toInt16(value));
  }

  readBinUint16Arg(): number {
    return this.readBinaryNumber((value) => BinaryConverter.toUint16(value));
  }

  readBinInt32Arg(): number {
    return this.readBinaryNumber((value) => BinaryConverter.toInt32(value));
  }

  readBinUint32Arg(): number {
    return this.readBinaryNumber((value) => BinaryConverter.toUint32(value));
  }

  readBinFloatArg(): number {
    return this.readBinaryNumber((value) => BinaryConverter.toFloat(value));
  }

  readBinDoubleArg(): number {
    if (this.communicationManager === undefined) {
      throw new Error('CommunicationManager was not set for command.');
    }
    return this.readBinaryNumber((value) =>
      this.communicationManager?.boardType === BoardType.Bit16
        ? BinaryConverter.toFloat(value)
        : BinaryConverter.toDouble(value)
    );
  }

  readBinByteArg(): number {
    return this.readBinaryNumber((value) => BinaryConverter.toByte(value));
  }

  readBinStringArg(): string {
    if (this.next()) {
      this.dumped = true;
      return latin1Decode(BinaryConverter.escapedStringToBytes(this.current()));
    }
    return '';
  }

  readBinBoolArg(): boolean {
    return this.readBinByteArg() !== 0;
  }

  read(format: string): unknown[] {
    return this.parseFormat(format).map((token) => this.readToken(token));
  }

  private readInteger(min: number, max: number): number {
    if (this.next()) {
      const parsed = parseInteger(this.current(), min, max);
      if (parsed !== null) {
        this.dumped = true;
        return parsed;
      }
    }
    return 0;
  }

  private readBinaryNumber(reader: (value: string) => number | null): number {
    if (this.next()) {
      const parsed = reader(this.current());
      if (parsed !== null) {
        this.dumped = true;
        return parsed;
      }
    }
    return 0;
  }

  private current(): string {
    return this.cmdArgs[this.parameter] ?? '';
  }

  private parseFormat(format: string): FormatToken[] {
    const tokens: FormatToken[] = [];
    let last: FormatToken | undefined;
    for (let i = 0; i < format.length; i += 1) {
      const char = format[i] ?? '';
      if (char === '*') {
        if (i === format.length - 1) {
          if (last === undefined) {
            throw new Error("Trailing '*' has no preceding format code to repeat.");
          }
          const remaining = this.cmdArgs.length - (this.parameter + (this.dumped ? 1 : 0)) - tokens.length;
          for (let j = 0; j < Math.max(remaining, 0); j += 1) {
            tokens.push(last);
          }
          continue;
        }
        const token = { binary: true, code: format[i + 1] ?? '' };
        tokens.push(token);
        last = token;
        i += 1;
      } else {
        const token = { binary: false, code: char };
        tokens.push(token);
        last = token;
      }
    }
    return tokens;
  }

  private readToken(token: FormatToken): unknown {
    const textReaders: Record<string, () => unknown> = {
      i: () => this.readInt32Arg(),
      I: () => this.readUint32Arg(),
      h: () => this.readInt16Arg(),
      H: () => this.readUint16Arg(),
      b: () => this.readInt16Arg(),
      B: () => this.readUint16Arg(),
      f: () => this.readFloatArg(),
      d: () => this.readDoubleArg(),
      s: () => this.readStringArg(),
      '?': () => this.readBoolArg(),
      c: () => this.readCharArg()
    };
    const binaryReaders: Record<string, () => unknown> = {
      i: () => this.readBinInt32Arg(),
      I: () => this.readBinUint32Arg(),
      h: () => this.readBinInt16Arg(),
      H: () => this.readBinUint16Arg(),
      b: () => this.readBinByteArg(),
      B: () => this.readBinByteArg(),
      f: () => this.readBinFloatArg(),
      d: () => this.readBinDoubleArg(),
      s: () => this.readBinStringArg(),
      '?': () => this.readBinBoolArg()
    };
    const reader = (token.binary ? binaryReaders : textReaders)[token.code];
    if (reader === undefined) {
      throw new Error(`Unknown format code ${token.binary ? '*' : ''}${token.code}`);
    }
    return reader();
  }
}
